algorithms/src/snark/marlin/marlin.rs

```diff
@@ -353,87 +349,89 @@ impl<
         Self::terminate(terminator)?;
 
         let eval_time = start_timer!(|| "Evaluating linear combinations over query set");
-        let mut evaluations_unsorted = Vec::new();
-        for (label, (_point_name, point)) in &query_set {
-            let lc =
-                lc_s.iter().find(|lc| &lc.label == label).ok_or_else(|| AHPError::MissingEval(label.to_string()))?;
-            let evaluation = polynomials.get_lc_eval(lc, *point)?;
-            if !AHPForR1CS::<TargetField, MM>::LC_WITH_ZERO_EVAL.contains(&lc.label.as_ref()) {
-                evaluations_unsorted.push((label.to_string(), evaluation));
+        let mut evaluations = std::collections::BTreeMap::new();
+        for (label, (_, point)) in query_set.to_set() {
+            if !AHPForR1CS::<E::Fr, MM>::LC_WITH_ZERO_EVAL.contains(&label.as_str()) {
+                let lc = lc_s
+                    .iter()
+                    .find(|lc| *lc.label == label)
```

Comment:
this looks like linear search; perhaps `lc_s` can be made into a `HashMap` for faster lookups, especially since these happen in a loop?

Reply:
Will do!